File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.java
class TrieNode 
{
    TrieNode[] children;
    boolean isWord;
    
    public TrieNode()
    {
        this.children = new TrieNode[26];
        this.isWord = false;
    }
}


class Trie {
    TrieNode root;

    public Trie() {
        root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode node = root;
        for(int i = 0; i < word.length(); i++)
        {
            char c = word.charAt(i);
            if(node.children[c - 'a'] == null)
            {
                node.children[c - 'a'] = new TrieNode();
            }
            node = node.children[c - 'a'];
        }
        
        node.isWord = true;
    }
    
    public boolean search(String word) {
        TrieNode node = root;
        for(int i = 0; i < word.length(); i++)
        {
            char c = word.charAt(i);
            if(node.children[c - 'a'] == null)
            {
                return false;
            }
            node = node.children[c - 'a'];
        }
        
        return node.isWord;
    }
    
    public boolean startsWith(String prefix) {
        TrieNode node = root;
        for(int i = 0; i < prefix.length(); i++)
        {
            char c = prefix.charAt(i);
            if(node.children[c - 'a'] == null)
            {
                return false;
            }
            node = node.children[c - 'a'];
        }
        
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.cpp
class TrieNode{
public:
    bool isWord;
    TrieNode* next[26];
    
    TrieNode()
    {
        this->isWord = false;
        for(int i = 0; i < 26; i++)
        {
            this->next[i] = nullptr;
        }
    }
};

class Trie {
public:
    TrieNode* root = new TrieNode();
    Trie() {
        
    }
    
    void insert(string word) {
        TrieNode* cursor = root;
        
        for(auto & c : word)
        {
            if(cursor->next[c - 'a'] == nullptr)
            {
                cursor->next[c - 'a'] = new TrieNode();
            }
            
            cursor = cursor->next[c - 'a'];
        }
        cursor->isWord = true;
    }
    
    bool search(string word) {
        TrieNode* cursor = root;
        
        for(auto & c : word)
        {
            if(cursor->next[c - 'a'] == nullptr)
            {
                return false;
            }
            
            cursor = cursor->next[c - 'a'];
        }
        
        return cursor->isWord;
    }
    
    bool startsWith(string prefix) {
        TrieNode* cursor = root;
        
        for(auto & c : prefix)
        {
            if(cursor->next[c - 'a'] == nullptr)
            {
                return false;
            }
            
            cursor = cursor->next[c - 'a'];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */